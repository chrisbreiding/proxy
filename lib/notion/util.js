const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
