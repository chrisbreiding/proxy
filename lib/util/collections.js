const compact = (array) => {
  return array.reduce((memo, item) => {
    return item ? memo.concat(item) : memo
  }, [])
}

module.exports = {
  compact,
}
