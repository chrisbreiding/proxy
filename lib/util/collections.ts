export function chunk<T = any> (array: T[], size: number) {
  if (!array.length) return []

  let index = 0
  const result = []

  while (index < array.length) {
    result.push(array.slice(index, index + size))

    index += size
  }

  return result
}

export function clone<T> (obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}

export function compact<T> (array: (T | undefined)[]): T[] {
  return array.reduce((memo, item) => {
    return item ? memo.concat(item) : memo
  }, [] as T[])
}

export async function mapPromisesSerially<T, U> (array: U[], callback: (arg: U) => Promise<T>) {
  const result: T[] = []

  for (const item of array) {
    result.push(await callback(item))
  }

  return result
}

export function times (num: number, filler?: any) {
  return (new Array(num)).fill(filler)
}
