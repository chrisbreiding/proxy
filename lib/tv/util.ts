import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
