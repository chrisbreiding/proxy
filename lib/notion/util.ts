import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
