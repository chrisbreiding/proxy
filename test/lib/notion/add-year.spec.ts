import nock from 'nock'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

process.env.API_KEY = 'key'

import { compact, times } from '../../../lib/util/collections'
import {
  nockGetBlockChildren,
  notionFixtureContents,
  block,
  snapshotAppendChildren,
  toQueryString,
  nockAppendBlockChildren,
} from './util'
import { handleServer } from '../../util'
import { startServer } from '../../..'

function makeBlocks (prefix: string, toggle = false, empty = false) {
  return {
    object: 'list',
    next_cursor: null,
    has_more: false,
    type: 'block',
    block: {},
    results: compact([
      block.bullet({ text: `${prefix} task` }),
      empty ? block.bullet({ text: '' }) : undefined,
      toggle ? block.toggle({ text: `${prefix} task w/ toggle`, id: 'blocks-from-toggle-id' }) : undefined,
    ]),
  }
}

function nockMost () {
  nockGetBlockChildren('pattern-month-id', { reply: makeBlocks('pattern-month-id') })
  nockGetBlockChildren('pattern-multiple-months-id', { reply: makeBlocks('pattern-multiple-months-id') })
  nockGetBlockChildren('pattern-multiple-months-date-id', { reply: makeBlocks('pattern-multiple-months-date-id') })
  nockGetBlockChildren('pattern-odd-months-id', { reply: makeBlocks('pattern-odd-months-id') })
  nockGetBlockChildren('pattern-odd-months-date-id', { reply: makeBlocks('pattern-odd-months-date-id') })
  nockGetBlockChildren('blocks-from-nested-id', { reply: makeBlocks('blocks-from-nested-id') })
  nockGetBlockChildren('blocks-from-toggle-id', { reply: makeBlocks('blocks-from-toggle-id') })
}

function nockOthers () {
  nockGetBlockChildren('pattern-month-date-id', { reply: makeBlocks('pattern-month-date-id', true, true) })
  nockGetBlockChildren('pattern-even-months-id', { reply: makeBlocks('pattern-even-months-id') })
  nockGetBlockChildren('pattern-even-months-date-id', { reply: makeBlocks('pattern-even-months-date-id') })
  nockGetBlockChildren('pattern-every-month-id', { reply: makeBlocks('pattern-every-month-id') })
  nockGetBlockChildren('pattern-every-month-date-id', { reply: makeBlocks('pattern-every-month-date-id') })
}

function makeQuery (updates: Record<string, string | null> = {}) {
  return toQueryString({
    action: 'addNextYear',
    notionToken: 'notion-token',
    futurePageId: 'future-page-id',
    year: '2023',
    ...updates,
  })
}

describe('lib/notion/add-year', () => {
  handleServer(startServer)

  beforeEach(() => {
    nock.cleanAll()

    vi.useFakeTimers()
    vi.setSystemTime(new Date(2022, 11, 28))

    nockMost()
    nockOthers()
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  it('adds blocks based on the year template patterns and year extras', async (ctx) => {
    nockGetBlockChildren('future-page-id', { fixture: 'add-year/future-blocks' })
    nockGetBlockChildren('year-template-id', { fixture: 'add-year/year-template-blocks' })
    nockGetBlockChildren('extras-id', { fixture: 'add-year/extras-blocks' })

    const snapshots = [
      snapshotAppendChildren({
        id: 'future-page-id',
        after: 'first-block-id',
        reply: { results: [
          block.p({ id: 'first-block-id' }),
          ...times(100).map((_, i) => block.bullet({ id: `block-id-${i + 1}` })),
        ] },
      }),
      snapshotAppendChildren({
        id: 'future-page-id',
        after: 'block-id-100',
        reply: { results: [
          block.p({ id: 'block-id-100' }),
          ...times(28, block.bullet()),
        ] },
      }),
    ]

    const res = await ctx.request.get(`/notion/action/key?${makeQuery({ year: null })}`)

    expect(res.status).to.equal(200)
    expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
    expect(res.text).to.include('Year 2023 successfully added!')

    await Promise.all(snapshots)
  })

  it('uses following year if not specified', async (ctx) => {
    nockGetBlockChildren('future-page-id', { reply: { results: [
      block.p({ id: 'first-block-id' }),
      block({ id: 'year-template-id', type: 'child_page', content: { title: 'Year Template' }, hasChildren: true }),
      block({ id: 'extras-id', type: 'child_page', content: { title: '2023' }, hasChildren: true }),
    ] } })
    nockGetBlockChildren('year-template-id', { reply: { results: [] } })
    nockGetBlockChildren('extras-id', { reply: { results: [] } })

    nockAppendBlockChildren({
      id: 'future-page-id',
    })

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.status).to.equal(200)
    expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
    expect(res.text).to.include('Year 2023 successfully added!')
  })

  it('handles deeply nested quests', async (ctx) => {
    nockGetBlockChildren('future-page-id', { reply: { results: [
      block.p({ id: 'first-block-id' }),
      block({ id: 'year-template-id', type: 'child_page', content: { title: 'Year Template' }, hasChildren: true }),
    ] } })
    nockGetBlockChildren('year-template-id', { reply: { results: [
      block.p({ id: 'month-id', text: 'January', hasChildren: true }),
    ] } })
    nockGetBlockChildren('month-id', { reply: { results: [
      block.bullet({ id: 'quest-id', text: 'Quest', hasChildren: true }),
    ] } })
    nockGetBlockChildren('quest-id', { reply: { results: [
      block.bullet({ id: 'nested-1-id', text: 'Nested 1', hasChildren: true }),
    ] } })
    nockGetBlockChildren('nested-1-id', { reply: { results: [
      block.bullet({ id: 'nested-2-id', text: 'Nested 2', hasChildren: true }),
    ] } })
    nockGetBlockChildren('nested-2-id', { reply: { results: [
      block.bullet({ text: 'Nested 3' }),
    ] } })

    const snapshots = [
      snapshotAppendChildren({
        after: 'first-block-id',
        id: 'future-page-id',
        reply: { results: [
          { id: 'first-block-id' },
          { id: '' }, { id: '' }, { id: '' },
          { id: 'appended-1-id' },
        ] },
      }),
      snapshotAppendChildren({
        id: 'appended-1-id',
        reply: { results: [{ id: 'appended-2-id' }] },
      }),
      snapshotAppendChildren({
        id: 'appended-2-id',
        reply: { results: [{ id: 'appended-3-id' }] },
      }),
      snapshotAppendChildren({
        id: 'appended-3-id',
        reply: { results: [{}] },
      }),
    ]

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.status).to.equal(200)
    expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
    expect(res.text).to.include('Year 2023 successfully added!')

    await Promise.all(snapshots)
  })

  // TODO: maybe not necesary and can remove?
  it.skip('handles adding more than 100 blocks', async (ctx) => {
    nockGetBlockChildren('future-page-id', { reply: { results: [
      block.p({ id: 'first-block-id' }),
      block({ id: 'year-template-id', type: 'child_page', content: { title: 'Year Template' }, hasChildren: true }),
      block({ id: 'extras-id', type: 'child_page', content: { title: '2023' }, hasChildren: true }),
    ] } })
    nockGetBlockChildren('year-template-id', { reply: { results: [] } })
    nockGetBlockChildren('extras-id', { reply: { results: [
      block.p({ text: 'January' }),
      block.p({ text: '1' }),
      ...times(120).map((_, i) => block.bullet({ text: `Block ${i + 1}` })),
    ] } })

    const snapshots = [
      snapshotAppendChildren({
        id: 'future-page-id',
        after: 'first-block-id',
        reply: { results: [
          block.p({ id: 'first-block-id' }),
          ...times(100).map((_, i) => block.bullet({ id: `block-id-${i + 1}` })),
        ] },
      }),
      snapshotAppendChildren({
        id: 'future-page-id',
        after: 'block-id-100',
        reply: { results: [
          block.p({ id: 'block-id-100' }),
          ...times(23, block.bullet()),
        ] },
      }),
    ]

    const res = await ctx.request.get(`/notion/action/key?${makeQuery({ year: null })}`)

    expect(res.status).to.equal(200)
    expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
    expect(res.text).to.include('Year 2023 successfully added!')

    await Promise.all(snapshots)
  })

  it('handles months that have no quests and lack of extras', async (ctx) => {
    nock.cleanAll()

    nockMost()

    const yearTemplateBlocks = notionFixtureContents('add-year/year-template-blocks')
    // remove blocks that would cause all months to have quests
    yearTemplateBlocks.results = [
      yearTemplateBlocks.results[0],
      ...yearTemplateBlocks.results.slice(2, 4),
    ]

    const futureBlocks = notionFixtureContents('add-year/future-blocks')
    // remove extras (2023) block
    futureBlocks.results = futureBlocks.results.filter((block: any) => {
      return block.child_page?.title !== '2023'
    })
    nockGetBlockChildren('future-page-id', { reply: futureBlocks })
    nockGetBlockChildren('year-template-id', { reply: yearTemplateBlocks })

    const snapshot = snapshotAppendChildren({
      id: 'future-page-id',
      after: 'first-block-id',
      reply: { results: times(23, block.bullet()) },
    })

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.status).to.equal(200)
    expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
    expect(res.text).to.include('Year 2023 successfully added!')

    await snapshot
  })

  it('sends 400 with error if futurePageId is not specified', async (ctx) => {
    const res = await ctx.request.get(`/notion/action/key?${makeQuery({ futurePageId: null })}`)

    expect(res.text).to.include('A value for <em>futurePageId</em> must be provided in the query string')
    expect(res.status).to.equal(400)
  })

  it('sends 400 with error if notionToken is not specified', async (ctx) => {
    const res = await ctx.request.get(`/notion/action/key?${makeQuery({ notionToken: null })}`)

    expect(res.text).to.include('A value for <em>notionToken</em> must be provided in the query string')
    expect(res.status).to.equal(400)
  })

  it('send 500 with error if year template cannot be found', async (ctx) => {
    const futureBlocks = notionFixtureContents('add-year/future-blocks')
    // remove year template block
    futureBlocks.results = futureBlocks.results.filter((block: any) => {
      return block.child_page?.title !== 'Year Template'
    })

    nockGetBlockChildren('future-page-id', { reply: futureBlocks })

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.text).to.include('Could not find year template')
    expect(res.status).to.equal(500)
  })

  it('sends 500 with error if extras date is not under a month', async (ctx) => {
    nockGetBlockChildren('future-page-id', { fixture: 'add-year/future-blocks' })
    nockGetBlockChildren('year-template-id', { fixture: 'add-year/year-template-blocks' })

    const extrasBlocks = notionFixtureContents('add-year/extras-blocks')
    // remove the first month block
    extrasBlocks.results = [
      extrasBlocks.results[0],
      ...extrasBlocks.results.slice(2),
    ]

    nockGetBlockChildren('extras-id', { reply: extrasBlocks })

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.text).to.include('Tried to add the following date, but could not determine the month for: \'1\'')
    expect(res.status).to.equal(500)
  })

  it('sends 500 with error if extras quest is not under a month', async (ctx) => {
    nockGetBlockChildren('future-page-id', { fixture: 'add-year/future-blocks' })
    nockGetBlockChildren('year-template-id', { fixture: 'add-year/year-template-blocks' })

    const extrasBlocks = notionFixtureContents('add-year/extras-blocks')
    // remove the first month block and date block
    extrasBlocks.results = [
      extrasBlocks.results[0],
      ...extrasBlocks.results.slice(3),
    ]

    nockGetBlockChildren('extras-id', { reply: extrasBlocks })

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.text).to.include('Tried to add the following quest, but could not determine the month for: \'Some task\'')
    expect(res.status).to.equal(500)
  })

  it('sends 500 with error if extras quest is not under a date', async (ctx) => {
    nockGetBlockChildren('future-page-id', { fixture: 'add-year/future-blocks' })
    nockGetBlockChildren('year-template-id', { fixture: 'add-year/year-template-blocks' })

    const extrasBlocks = notionFixtureContents('add-year/extras-blocks')
    // remove the first date block
    extrasBlocks.results = [
      ...extrasBlocks.results.slice(0, 2),
      ...extrasBlocks.results.slice(3),
    ]

    nockGetBlockChildren('extras-id', { reply: extrasBlocks })

    const res = await ctx.request.get(`/notion/action/key?${makeQuery()}`)

    expect(res.text).to.include('Tried to add the following quest, but could not determine the date for: \'Some task\'')
    expect(res.status).to.equal(500)
  })
})
