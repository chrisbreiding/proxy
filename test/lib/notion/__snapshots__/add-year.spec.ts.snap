// Vitest Snapshot v1

exports[`lib/notion/add-year > adds blocks based on the year template patterns and year extras 1`] = `
"
**January**
Sun, 1/1
- pattern-month-id task
- pattern-multiple-months-id task
- pattern-odd-months-id task
- pattern-every-month-id task
- Some task
- Some event
Mon, 1/2
- Another event
Thu, 1/5
- pattern-odd-months-date-id task
Tue, 1/10
- pattern-every-month-date-id task
Fri, 1/20
- pattern-multiple-months-date-id task
Tue, 1/24
- pattern-month-date-id task
> pattern-month-date-id task w/ toggle
	- blocks-from-toggle-id task

**February**
Wed, 2/1
- pattern-even-months-id task
- pattern-every-month-id task
Fri, 2/10
- pattern-every-month-date-id task
Wed, 2/22
- Another task
Thu, 2/23
- pattern-even-months-date-id task

**March**
Wed, 3/1
- pattern-odd-months-id task
- pattern-every-month-id task
- Has a subtask
	- blocks-from-nested-id task
- Event after subtask
Sun, 3/5
- pattern-odd-months-date-id task
Fri, 3/10
- pattern-every-month-date-id task

**April**
Sat, 4/1
- pattern-multiple-months-id task
- pattern-even-months-id task
- pattern-every-month-id task
Mon, 4/10
- pattern-every-month-date-id task
Sun, 4/23
- pattern-even-months-date-id task

**May**
Mon, 5/1
- pattern-odd-months-id task
- pattern-every-month-id task
Fri, 5/5
- pattern-odd-months-date-id task
Wed, 5/10
- pattern-every-month-date-id task

**June**
Thu, 6/1
- pattern-even-months-id task
- pattern-every-month-id task
Sat, 6/10
- pattern-every-month-date-id task
Fri, 6/23
- pattern-even-months-date-id task

**July**
Sat, 7/1
- pattern-multiple-months-id task
- pattern-odd-months-id task
- pattern-every-month-id task
Wed, 7/5
- pattern-odd-months-date-id task
Mon, 7/10
- pattern-every-month-date-id task

**August**
Tue, 8/1
- pattern-even-months-id task
- pattern-every-month-id task
Thu, 8/10
- pattern-every-month-date-id task
Wed, 8/23
- pattern-even-months-date-id task

**September**
Fri, 9/1
- pattern-odd-months-id task
- pattern-every-month-id task
Tue, 9/5
- pattern-odd-months-date-id task
Sun, 9/10
- pattern-every-month-date-id task
Wed, 9/20
- pattern-multiple-months-date-id task
"
`;

exports[`lib/notion/add-year > adds blocks based on the year template patterns and year extras 2`] = `
"
**October**
Sun, 10/1
- pattern-multiple-months-id task
- pattern-even-months-id task
- pattern-every-month-id task
Tue, 10/10
- pattern-every-month-date-id task
Mon, 10/23
- pattern-even-months-date-id task

**November**
Wed, 11/1
- pattern-odd-months-id task
- pattern-every-month-id task
Sun, 11/5
- pattern-odd-months-date-id task
Fri, 11/10
- pattern-every-month-date-id task

**December**
Fri, 12/1
- pattern-even-months-id task
- pattern-every-month-id task
Sun, 12/10
- pattern-every-month-date-id task
Sat, 12/23
- pattern-even-months-date-id task
"
`;

exports[`lib/notion/add-year > handles deeply nested quests 1`] = `
"
**January**
Sun, 1/1
- Quest
"
`;

exports[`lib/notion/add-year > handles deeply nested quests 2`] = `
"- Nested 1
"
`;

exports[`lib/notion/add-year > handles deeply nested quests 3`] = `
"- Nested 2
"
`;

exports[`lib/notion/add-year > handles deeply nested quests 4`] = `
"- Nested 3
"
`;

exports[`lib/notion/add-year > handles months that have no quests and lack of extras 1`] = `
"
**January**
Sun, 1/1
- pattern-month-id task
- pattern-multiple-months-id task
Fri, 1/20
- pattern-multiple-months-date-id task

**April**
Sat, 4/1
- pattern-multiple-months-id task

**July**
Sat, 7/1
- pattern-multiple-months-id task

**September**
Wed, 9/20
- pattern-multiple-months-date-id task

**October**
Sun, 10/1
- pattern-multiple-months-id task
"
`;
